### DISCUSSION TITLE
checking permissions with new AIR + new Android

### DISCUSSION THREAD

**Discussion by stephanberger4** [airsdk/Adobe-Runtime-Support#1566] [Q&A]
Hi, 

I've got an app since 5 years that requires storage access to download and display an offline map. 
Since the new AIR release (and probably since the new Android release), my app is refusing to display the map : 
`Error #3001: File or directory access denied.`
I didn't change the code of my app, so I suppose it's because of the [new version of Android](https://developer.android.com/about/versions/11/privacy/storage)

I was using the ANE "[PermissionCheck](https://github.com/myflashlab/PermissionCheck-ANE)" of myflashlab but the ANE hasn't been updated since 2 years old (and myflashlab closed...).

My current code is : 

```
Function checkPermissions():void{
		trace("Checking Permissions");
PermissionCheck.init();	
var permissionState:int = PermissionCheck.check(PermissionCheck.SOURCE_STORAGE);
	
		if (permissionState == PermissionCheck.PERMISSION_UNKNOWN || permissionState == PermissionCheck.PERMISSION_DENIED)
		{
			PermissionCheck.request(PermissionCheck.SOURCE_STORAGE, onRequestResult);
			trace("Request Permission Storage");
		}
		else
		{
			trace("permission storage ok");
			requestGeolocationPermission();
		}

		function onRequestResult($state:int):void
		{
			if ($state != PermissionCheck.PERMISSION_GRANTED)
			{
				trace("STORAGE NOT ALLOWED");
				

			}
			else
			{
				//init();
				trace("STORAGE is allowed by user");
			}
		}
	}
```

And my XML has : 

```
<android>

        <colorDepth>16bit</colorDepth>

        <manifestAdditions><![CDATA[

<manifest android:installLocation="auto">
                 <uses-sdk android:minSdkVersion="19" android:targetSdkVersion="30" />
android:requestLegacyExternalStorage="true"
                 <uses-permission android:name="android.permission.INTERNET"/>
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"
    android:maxSdkVersion="28" />
<uses-permission android:name="android.permission.MANAGE_EXTERNAL_STORAGE" />
```

Can someone help me to update my code for checking Permissions in AS3 with these new restrictions please ? 

It would help a lot ! 

Thank you,

**Comment by 2jfw** [comment]
Related: https://github.com/airsdk/Adobe-Runtime-Support/issues/939

**Comment by ajwfrost** [comment]
Hi

I don't think the `requestLegacyExternalStorage` flag will do anything here I'm afraid (your application is targeting API 30 = Android 11)
https://developer.android.com/about/versions/11/privacy/storage

So mostly I imagine the issue is that you're trying to create a file in the 'root' folder of the external storage directory which is provided as `File.documentsDirectory` ... can you try moving your content so that it's instead put in one of the subfolders, e.g.
`File.documentsDirectory.resolvePath("Documents")`

thanks

**Reply by stephanberger4** [reply]
Thank you very much, I'll try that and let you know !

**Reply by stephanberger4** [reply]
Ok. I've tried but I'm stuck...
It seems that Android (or AIR) won't "allow" to copy files from `applicationStorage `to `Documents`. 
I've got an "assets" folder embed to my app that I'm trying to copy on "Documents" to be able to access it anytime without restrictions (Android 11). 

What I did : 
```
source = File.applicationDirectory.resolvePath("assets/maps");
destination = File.documentsDirectory.resolvePath("Documents/www");
source.addEventListener(Event.COMPLETE, onMapCopyComplete);
source.copyToAsync(destination, false);

function onMapCopyComplete(e: Event): void {
	source.removeEventListener(Event.COMPLETE, onMapCopyComplete);

	log("onMapCopyComplete()");
}
```

I've got a return `onMapCopyComplete()` but when I'm looking in `InternalStorage/Documents` of my phone I've got the folders but all are empty... None of the files were copy..

![image](https://user-images.githubusercontent.com/6382334/149484181-113ea1af-9c7e-45b2-869c-2e0e0a28ecf5.png)

Any idea what's happening ?

**Reply by mesaglam** [reply]
Hello, were you able to fix this problem?

**Reply by pol2095** [reply]
You can use my ANE [Storage Access Framework](http://pol2095.free.fr/Android-Storage-Access-Framework/) ([documentation](http://pol2095.free.fr/Android-Storage-Access-Framework/docs/index.html)) but I don't know if it works for you, it's necessary to open the file explorer the first time to save a file, it's the new Android's behavior.
SAF doesn't need permission to work.

**Reply by mesaglam** [reply]
Thanks for your great effort, you rock!

**Comment by balaedit** [comment]
> You can use my ANE [Storage Access Framework](http://pol2095.free.fr/Android-Storage-Access-Framework/) ([documentation](http://pol2095.free.fr/Android-Storage-Access-Framework/docs/index.html)) but I don't know if it works for you, it's necessary to open the file explorer the first time to save a file, it's the new Android's behavior. SAF doesn't need permission to work.

@pol2095 may I get your email id or drop me your email at [balaedit@gmail.com](mailto:balaedit@gmail.com). I have to discuss regarding Storage Access Framework.

**Reply by pol2095** [reply]
ask your question here or a private message on [Starling forum](https://forum.starling-framework.org/)

**Reply by balaedit** [reply]
AIR SDK 50.2.3.1
@pol2095 Please share the sample file of saving and reading text file / binary file using SAF ANE (Android).
